db/pebbledb/pebledb.go

```diff
@@ -0,0 +1,158 @@
+
+// Get implements the db.ReadTx.Get interface method
+func (tx ReadTx) Get(k []byte) ([]byte, error) {
+	v, closer, err := tx.batch.Get(k)
+	if err == pebble.ErrNotFound {
+		return nil, db.ErrKeyNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	if err = closer.Close(); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// Discard implements the db.ReadTx.Discard interface method
+func (tx ReadTx) Discard() {
+	// Close returns an error, but here in the Discard context is ommited
+	tx.batch.Close()
+}
+
+// Get implements the db.WriteTx.Get interface method
+func (tx WriteTx) Get(k []byte) ([]byte, error) {
+	return ReadTx(tx).Get(k)
+}
+
+// Set implements the db.WriteTx.Set interface method
+func (tx WriteTx) Set(k, v []byte) error {
+	return tx.batch.Set(k[:], v, nil)
+}
+
+// Delete implements the db.WriteTx.Delete interface method
+func (tx WriteTx) Delete(k []byte) error {
+	return tx.batch.Delete(k, nil)
+}
+
+// Commit implements the db.WriteTx.Commit interface method
+func (tx WriteTx) Commit() error {
+	return tx.batch.Commit(nil)
+}
+
+// Discard implements the db.WriteTx.Discard interface method
+func (tx WriteTx) Discard() {
+	ReadTx(tx).Discard()
+}
+
+// PebbleDB implements db.Database interface
+type PebbleDB struct {
+	db *pebble.DB
+}
+
+// check that PebbleDB implements the db.Database interface
+var _ db.Database = (*PebbleDB)(nil)
+
+// Options defines params for creating a new PebbleDB
+type Options struct {
+	Path string
+}
+
+// New returns a PebbleDB using the given Options, which implements the
+// db.Database interface
+func New(opts Options) (*PebbleDB, error) {
+	if err := os.MkdirAll(opts.Path, os.ModePerm); err != nil {
+		return nil, err
+	}
+	o := &pebble.Options{}
+	db, err := pebble.Open(opts.Path, o)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PebbleDB{
+		db: db,
+	}, nil
+}
+
+// ReadTx returns a db.ReadTx
+func (db *PebbleDB) ReadTx() db.ReadTx {
+	return ReadTx{
+		batch: db.db.NewIndexedBatch(),
+	}
+}
+
+// WriteTx returns a db.WriteTx
+func (db *PebbleDB) WriteTx() db.WriteTx {
+	return WriteTx{
+		batch: db.db.NewIndexedBatch(),
+	}
+}
+
+// Close closes the PebbleDB
+func (db *PebbleDB) Close() error {
+	return db.db.Close()
+}
+
+func keyUpperBound(b []byte) []byte {
+	// https://github.com/cockroachdb/pebble/blob/b2eb88a7182687c81d911c425309ef0e1f545452/iterator_example_test.go#L44
+	end := make([]byte, len(b))
+	copy(end, b)
+	for i := len(end) - 1; i >= 0; i-- {
+		end[i] = end[i] + 1
+		if end[i] != 0 {
+			return end[:i+1]
+		}
+	}
+	return nil // no upper-bound
+}
+
+// Iterate implements the db.Database.Iterate interface method
+func (db *PebbleDB) Iterate(prefix []byte, callback func(k, v []byte) bool) error {
+	var err error
+	iterOptions := &pebble.IterOptions{
+		LowerBound: prefix,
+		UpperBound: keyUpperBound(prefix),
+	}
+	iter := db.db.NewIter(iterOptions)
+	defer func() {
+		errC := iter.Close()
+		if err != nil {
+			return
+		}
+		err = errC
```

Comment:
I don't think this error ends up anywhere, unless you use `) (err error) {` for the return parameter.

Reply:
Right! Updated.